### PR TITLE
Mob/storage caps

### PR DESCRIPTION
**Pull request by fabmiz**
- Ticket: [ENG-2256]
- Feature flag: n/a

## Purpose

Add the ability to notify users if they are approaching a node's storage cap
Prevent certain file operation that'd cause a node to exceed its storage cap

## Summary of Changes

- Add node-storage relationship to node, add node-storage model/serializer/adapter
- Mirage factory updates
- Add tests
- Modify quickfiles browser and files-widget to show insufficient storage API error messages
warn users if approaching storage cap.

## QA Notes

[ENG-2256]: https://openscience.atlassian.net/browse/ENG-2256

### CODE REVIEW

File: app/models/node-storage.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+export enum StorageStatus {
+    DEFAULT = 'DEFAULT',
+    APPROACHING_PRIVATE = 'APPROACHING_PRIVATE',
+    OVER_PRIVATE = 'OVER_PRIVATE',
+    APPROACHING_PUBLIC = 'APPROACHING_PUBLIC',
+    OVER_PUBLIC = 'OVER_PUBLIC',
+    NOT_CALCULATED = 'NOT_CALCULATED',
+}
+
+export default class NodeStorageModel extends OsfModel {
```
Comment (fabmiz) [inline app/models/node-storage.ts:17]: Add model unit tests, test computed props

File: lib/osf-components/addon/components/file-browser/component.ts [hunk 2]
```diff
@@ -178,6 +179,20 @@ export default class FileBrowser extends Component {
         ].filter(item => item.length);
     }
 
+    @computed('node.public', 'isProjectSelectorValid', 'isMoving')
+    get moveAllowed(): boolean {
```
Comment (fabmiz) [inline lib/osf-components/addon/components/file-browser/component.ts:183]: Rename to something more specific. `should...`;

File: lib/osf-components/addon/components/file-browser/file-move-messages/component.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import NodeStorageModel from 'ember-osf-web/models/node-storage';
+
+import template from './template';
+
+@layout(template)
+export default class FileBrowserFileMoveMessages extends Component {
+    project!: Node;
+
+    @alias('project.public') isPublicProject!: boolean;
+    @bool('project.links.relationships.parent') isChildNode!: boolean;
```
Comment (fabmiz) [inline lib/osf-components/addon/components/file-browser/file-move-messages/component.ts:16]: maybe `project.isRoot` for better readability

File: lib/osf-components/addon/components/file-browser/file-move-messages/component.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+    @computed('project.storage')
+    get selectedStorageStatus() {
```
Comment (fabmiz) [inline lib/osf-components/addon/components/file-browser/file-move-messages/component.ts:19]:
```suggestion
    get selectedProjectStorageStatus() {
```

File: app/guid-user/quickfiles/controller.ts [hunk 5]
```diff
@@ -107,13 +107,17 @@ export default class UserQuickfiles extends Controller {
...
-
             yield file.move(node);
             yield this.flash.perform(file, this.intl.t('file_browser.successfully_moved'));
             this.allFiles.removeObject(file);
             return true;
-        } catch (ex) {
-            this.toast.error(this.intl.t('move_to_project.could_not_move_file'));
+        } catch (e) {
+            let toastMessage = this.intl.t('move_to_project.could_not_move_file');
+            if (e.status === '507') {
```
Comment (fabmiz) [inline app/guid-user/quickfiles/controller.ts:116]:
```suggestion
            if (e.status === 507) {
```

File: mirage/factories/node-storage.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    overPrivate: trait<NodeStorageModel>({
+        afterCreate(nodeStorage) {
+            nodeStorage.update('storageLimitStatus', StorageStatus.OVER_PRIVATE);
+        },
+    }),
+    overPublic: trait<NodeStorageModel>({
+        afterCreate(nodeStorage) {
+            nodeStorage.update('storageLimitStatus', StorageStatus.OVER_PUBLIC);
+        },
+    }),
```
Comment (fabmiz) [inline mirage/factories/node-storage.ts:37]: Add `NOT_CALCULATED` trait.

File: mirage/views/file.ts [hunk 7]
```diff
@@ -45,6 +46,13 @@ export function uploadToRoot(this: HandlerContext, schema: Schema) {
     const randomNum = faker.random.number();
     const fileGuid = guid('file');
     const id = fileGuid(randomNum);
+
+    if (!node.public && node.storage.storageLimitStatus === StorageStatus.OVER_PRIVATE) {
```
Comment (futa-ikeda) [inline mirage/views/file.ts:50]: use storage limit computeds here

File: tests/acceptance/guid-user/quickfiles-test.ts [hunk 8]
```diff
@@ -154,6 +154,8 @@ module('Acceptance | Guid User Quickfiles', hooks => {
                     public: true,
                 },
             );
+            const nodeStorage = server.create('node-storage');
```
Comment (futa-ikeda) [inline tests/acceptance/guid-user/quickfiles-test.ts:157]: Use `withStorage` trait on node

File: tests/acceptance/guid-user/quickfiles-test.ts [hunk 9]
```diff
@@ -195,6 +197,8 @@ module('Acceptance | Guid User Quickfiles', hooks => {
                     public: false,
                 },
             );
+            const nodeStorage = server.create('node-storage');
+            node.update({ storage: nodeStorage });
```
Comment (futa-ikeda) [inline tests/acceptance/guid-user/quickfiles-test.ts:201]: same here

File: tests/acceptance/guid-user/quickfiles-test.ts [hunk 10]
```diff
@@ -239,6 +243,8 @@ module('Acceptance | Guid User Quickfiles', hooks => {
                 'node',
                 { title, lastLogged: '2017-10-19T12:05:10.571Z', dateModified: '2017-10-19T12:05:10.571Z' },
             );
+            const nodeStorage = server.create('node-storage');
+            node.update({ storage: nodeStorage });
```
Comment (futa-ikeda) [inline tests/acceptance/guid-user/quickfiles-test.ts:247]: and here

File: tests/acceptance/guid-user/quickfiles-test.ts [hunk 11]
```diff
@@ -275,6 +281,8 @@ module('Acceptance | Guid User Quickfiles', hooks => {
...
```
Comment (futa-ikeda) [inline tests/acceptance/guid-user/quickfiles-test.ts:285]: and here too

**Review by futa-ikeda — CHANGES_REQUESTED** [review]

File: app/models/node-storage.ts [hunk 12]
```diff
@@ -5,6 +5,61 @@ import OsfModel from './osf-model';
 
 const { attr } = DS;
 
+export const TestCases: Record<'public' | 'private', Record<StorageStatus, { warning: boolean, error: boolean }>> = {
```
Comment (fabmiz) [inline app/models/node-storage.ts:8]: @futa-ikeda I don't think this is a good place for this const. It's not the model's concern and negatively affects readability. You could instead create a module that exports this const under `tests/fixtures/`

**Pull request merged**